### PR TITLE
Inform players of new unit health through web socket

### PR DESCRIPTION
**Pull request by hilalcomak**
Solves #202

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java [hunk 1]
```diff
@@ -32,8 +43,20 @@ public class GameController {
...
+            String endPosition = asJsonString(unitCommandPutDTO.getEnd());
+            String defenderPosition = asJsonString(DTOMapper.INSTANCE.convertPositionToPositionDTO(defendingUnit.getPosition()));
+            String payload = String.format("{\"start\": %s, \"end\": %s, \"defenderPosition\":%s, \"health\":%s}",
+                    startPosition, endPosition, defenderPosition, defendingUnit.getHealth());
+            socketMessage.convertAndSend(TOPIC_GAME + id, payload);
+        }
+        catch (JsonProcessingException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+                    String.format("Message body could not be sent: %s", e));
+        }
```
[the remark this answers is not in the document]
Comment (paolovic) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java:59]: Luis is right imo

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java [hunk 2]
```diff
@@ -32,8 +43,20 @@ public class GameController {
     public void unitAttack(@RequestHeader("token") String token, @PathVariable Long id, @RequestBody UnitCommandPutDTO unitCommandPutDTO) {
         Position attacker = DTOMapper.INSTANCE.convertPositionDTOToPosition(unitCommandPutDTO.getStart());
         Position defender = DTOMapper.INSTANCE.convertPositionDTOToPosition(unitCommandPutDTO.getEnd());
+        Unit defendingUnit = gameService.unitAttack(id, token, attacker, defender);
 
-        gameService.unitAttack(id, token, attacker, defender);
+        try {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java:48]: we shouldn't "try and catch" in a controller

if you do it like @luis-tm below proposed we won't need that

**Review by paolovic — CHANGES_REQUESTED** [review]

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java (hunk 2 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java:48]: @luis-tm way looks cleaner, but why should't we try/catch in controller?

Now it is is just hidden in the framework (https://github.com/spring-projects/spring-framework/blob/061d1263f854a132d76f3528034f907107c1727e/spring-messaging/src/main/java/org/springframework/messaging/core/AbstractMessageSendingTemplate.java#L186)

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java [hunk 3]
```diff
@@ -43,5 +66,20 @@ public void unitWait(@RequestHeader("token") String token, @PathVariable Long id
...
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+                    String.format("Message body could not be sent: %s", e));
+        }
     }
+
+    private static String asJsonString(final Object object) throws JsonProcessingException {
+        return new ObjectMapper().writeValueAsString(object);
+    }
+
 }
```
[the remark this answers is not in the document]
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java]: done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/HealthDTO.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package ch.uzh.ifi.hase.soprafs22.rest.dto;
+
+public class HealthDTO {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/HealthDTO.java:3]: Can we please go back to where we stated what kind of DTO it is?
Like HealthGetDTO or HealthPutDTO?

I know, I said anytime anywhere that I like it but this was not a well thought through idea from my side, sorry for that.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/MovementDTO.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package ch.uzh.ifi.hase.soprafs22.rest.dto;
+
+public class MovementDTO {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/MovementDTO.java:3]: same

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UnitDeltaSockDTO.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package ch.uzh.ifi.hase.soprafs22.rest.dto;
+
+public class UnitDeltaSockDTO {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UnitDeltaSockDTO.java:3]: UnitDeltaSock? :D

Bildschirmfoto 2022-05-08 um 12 32 53

What's the purpose of this class? We need it when for example a unit was killed and we want to tell the client about that over WebSocket, right?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/HealthDTO.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    private PositionDTO defenderPosition;
+    private long health;
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/HealthDTO.java:5]: why the upcasting to a long?
usually our health is an int, that should be enough in my oppinion

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UnitDeltaSockDTO.java (hunk 6 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UnitDeltaSockDTO.java:3]: :D

Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UnitDeltaSockDTO.java:3]: delta

**Pull request merged**